### PR TITLE
Separate public and private headers

### PR DESCRIPTION
**Pull request by NikitaZotov**
* [x] Read PR [documentation](https://github.com/ostis-ai/sc-machine/blob/main/CONTRIBUTING.md)
* [x] Update changelog
* [x] Update documentation

This PR unblocks tasks for build system improvement.

Related tasks:

- Remove `configure_file` for tests and use `target_compile_options`

Links:

https://www.reddit.com/r/cpp_questions/comments/m7k0bs/project_structure_in_c/?rdt=43314
https://slaptijack.com/programming/project-structure-for-cpp.html
https://www.linkedin.com/pulse/what-general-c-project-structure-like-herbert-elwood-gilliland-iii/
https://hiltmon.com/blog/2013/07/03/a-simple-c-plus-plus-project-structure/

### CODE REVIEW

**Review by KovalM — APPROVED** [review]

**Review by KovalM — APPROVED** [review]

File: sc-memory/sc-memory/src/utils/sc_message.cpp [hunk 1]
```diff
@@ -1,4 +1,11 @@
-#include "sc_message.hpp"
+/*
+ * This source file is part of an OSTIS project. For the latest info, see http://ostis.net
+ * Distributed under the MIT License
+ * (See accompanying file COPYING.MIT or copy at http://opensource.org/licenses/MIT)
+ */
+
+#include "sc-memory/utils/sc_message.hpp"
+
 #include "../sc_platform.hpp"
```
Comment (kilativ-dotcom) [inline sc-memory/sc-memory/src/utils/sc_message.cpp:9]: is .. in path ok?

**Review by kilativ-dotcom — CHANGES_REQUESTED** [review]

File: sc-kpm/sc-agents-common/src/GenerationUtils.cpp [hunk 2]
```diff
@@ -4,12 +4,12 @@
...
-#include <algorithm>
+#include "sc-agents-common/utils/GenerationUtils.hpp"
 
-#include "GenerationUtils.hpp"
+#include <algorithm>
 
-#include "IteratorUtils.hpp"
-#include "CommonUtils.hpp"
+#include "sc-agents-common/utils/IteratorUtils.hpp"
+#include "sc-agents-common/utils/CommonUtils.hpp"
```
[the remark this answers is not in the document]
Comment (NikitaZotov) [reply sc-kpm/sc-agents-common/src/GenerationUtils.cpp:12]: What do you mean by different order?

File: sc-memory/sc-core/src/sc-store/sc-base/sc_monitor_private.h [hunk 3]
```diff
@@ -0,0 +1,27 @@
...
+#include "sc-core/sc-container/sc_queue.h"
+
+#include "sc_mutex_private.h"
+
+struct _sc_monitor
+{
+  sc_mutex rw_mutex;         // Mutex for data protection
+  sc_queue queue;            // Queue of writers and readers
+  sc_uint32 active_readers;  // Number of readers currently accessing the data
+  sc_uint32 active_writer;   // Flag to indicate if a writer is writing
```
[the remark this answers is not in the document]
Comment (NikitaZotov) [reply sc-memory/sc-core/src/sc-store/sc-base/sc_monitor_private.h:21]: Doesn't make any difference, there will still be structure alignment

File: sc-memory/sc-core/src/sc-store/sc_storage.h [hunk 4]
```diff
@@ -89,6 +89,10 @@ sc_bool sc_storage_is_initialized();
  */
 sc_bool sc_storage_is_element(sc_memory_context const * ctx, sc_addr addr);
 
+void sc_storage_start_new_process();
+
+void sc_storage_end_new_process();
```
[the remark this answers is not in the document]
Comment (NikitaZotov) [reply sc-memory/sc-core/src/sc-store/sc_storage.h:94]: No, I didn't change anything in the code, I just moved these methods around

File: sc-kpm/sc-agents-common/src/GenerationUtils.cpp (hunk 2 above)
[the remark this answers is not in the document]
Comment (kilativ-dotcom) [reply sc-kpm/sc-agents-common/src/GenerationUtils.cpp:12]: now utils are splitted by algorithm and not in alphabetical order

File: sc-kpm/sc-common/include/sc-common/sc_utils.h [hunk 5]
```diff
@@ -4,35 +4,37 @@
...
- * @param el sc-addr of sc-element to append into asnwer
+ * @param el sc-addr of sc-element to append into result
  */
-void appendIntoResult(sc_addr result, sc_addr el);
+void appendIntoResult(sc_memory_context * context, sc_addr result, sc_addr el);
 
 /*!
- * Remove action from action_initiated set and append it into
+ * Append action into action_finished set.
  * action_finished set.
```
Comment (kilativ-dotcom) [inline sc-kpm/sc-common/include/sc-common/sc_utils.h:34]:
```suggestion
```

File: sc-memory/sc-core/src/sc-store/sc_storage.h (hunk 4 above)
[the remark this answers is not in the document]
Comment (kilativ-dotcom) [reply sc-memory/sc-core/src/sc-store/sc_storage.h:94]: but they are in src folder and it is private

File: sc-kpm/sc-agents-common/src/GenerationUtils.cpp (hunk 2 above)
[the remark this answers is not in the document]
Comment (NikitaZotov) [reply sc-kpm/sc-agents-common/src/GenerationUtils.cpp:12]: Alphabetical order is optional

File: sc-memory/sc-memory/src/utils/sc_signal_handler.cpp [hunk 6]
```diff
@@ -4,10 +4,11 @@
...
  */
 
-#include "sc_signal_handler.hpp"
+#include "sc-memory/utils/sc_signal_handler.hpp"
 
-#include "../sc_platform.hpp"
-#include "../sc_debug.hpp"
+#include "sc-memory/sc_debug.hpp"
+
+#include "sc_platform.hpp"
```
Comment (kilativ-dotcom) [inline sc-memory/sc-memory/src/utils/sc_signal_handler.cpp:11]: sc-memory before sc_platform?

**Review by kilativ-dotcom — CHANGES_REQUESTED** [review]

File: sc-memory/sc-core/src/sc-store/sc_storage.h (hunk 4 above)
[the remark this answers is not in the document]
Comment (NikitaZotov) [reply sc-memory/sc-core/src/sc-store/sc_storage.h:94]: Yes, they are private. They don't have to be open, they reveal the features of the sc-memory

File: sc-memory/sc-memory/src/utils/sc_signal_handler.cpp (hunk 6 above)
Comment (NikitaZotov) [reply sc-memory/sc-memory/src/utils/sc_signal_handler.cpp:11]: It is in src

File: sc-kpm/sc-agents-common/src/GenerationUtils.cpp (hunk 2 above)
[the remark this answers is not in the document]
Comment (kilativ-dotcom) [reply sc-kpm/sc-agents-common/src/GenerationUtils.cpp:12]: won't you unsplit utils includes?

**Review by kilativ-dotcom — APPROVED** [review]
looks ok :+1:

**Pull request merged**